src/vector/text/vtextshaper.cpp

```diff
@@ -0,0 +1,244 @@
+        else if (!strcasecmp(fontWeight, "light")) return FC_WEIGHT_LIGHT;
+        else if (!strcasecmp(fontWeight, "bold")) return FC_WEIGHT_NORMAL;
+        else return FC_WEIGHT_NORMAL;
+    }
+    VFontConfig() {
+        mFontconfig = FcInitLoadConfigAndFonts();
+    }
+    ~VFontConfig() {
+        FcConfigDestroy(mFontconfig);
+    }
+private:
+    FcConfig *mFontconfig{nullptr};
+};
+
+class VFont
+{
+public:
+    VFont(FT_Library library, std::string fontPath, int size);
+    ~VFont();
+    VPath outline(FT_UInt glyphIndex);
+    FT_Face ftFace() const;
+private:
+#define FROM_FT_COORD(x) ((x) / 64.)
+#define MINUS(x) ((-1) * (x))
+    static int moveTo(const FT_Vector *to, void *user);
+    static int lineTo(const FT_Vector *to, void *user);
+    static int conicTo(const FT_Vector *control, const FT_Vector *to, void *user);
+    static int cubicTo(const FT_Vector *control1, const FT_Vector *control2,
+            const FT_Vector *to, void *user);
+private:
+    static thread_local FT_Face         mFace;
+    constexpr static FT_Outline_Funcs   mOutline_funcs = {
+        moveTo,
+        lineTo,
+        conicTo,
+        cubicTo,
+        0, 0
+    };
+    std::unordered_map<FT_UInt, VPath>  mMap;
+};
+
+thread_local FT_Face VFont::mFace = {};
+constexpr FT_Outline_Funcs VFont::mOutline_funcs;
+
+VFont::VFont(FT_Library library, std::string fontPath, int size) {
+    if (FT_New_Face(library, fontPath.c_str(), 0, &mFace)) {
+        std::cout << "Font file load failed :" << fontPath << std::endl;
+    }
+    FT_Set_Char_Size(mFace, size << 6, size << 6, 0, 0);
+}
+
+VFont::~VFont() {
+    FT_Done_Face(mFace);
+}
+
+FT_Face VFont::ftFace() const {
+    return mFace;
+}
+
+int VFont::moveTo(const FT_Vector *to, void *user)
+{
+    VPath *path = (VPath *)user;
+    path->moveTo(FROM_FT_COORD(to->x), FROM_FT_COORD(MINUS(to->y)));
+    return 0;
+}
+
+int VFont::lineTo(const FT_Vector *to, void *user)
+{
+    VPath *path = (VPath *)user;
+    path->lineTo(FROM_FT_COORD(to->x), FROM_FT_COORD(MINUS(to->y)));
+    return 0;
+}
+
+int VFont::conicTo(const FT_Vector *control, const FT_Vector *to, void *user)
+{
+    VPath *path = (VPath *)user;
+    path->cubicTo(
+            FROM_FT_COORD(control->x), FROM_FT_COORD(MINUS(control->y)),
+            FROM_FT_COORD(control->x), FROM_FT_COORD(MINUS(control->y)),
+            FROM_FT_COORD(to->x), FROM_FT_COORD(MINUS(to->y)));
+    return 0;
+}
+
+int VFont::cubicTo(const FT_Vector *control1, const FT_Vector *control2,
+        const FT_Vector *to, void *user)
+{
+    VPath *path = (VPath *)user;
+    path->cubicTo(
+            FROM_FT_COORD(control1->x), FROM_FT_COORD(MINUS(control1->y)),
+            FROM_FT_COORD(control2->x), FROM_FT_COORD(MINUS(control2->y)),
+            FROM_FT_COORD(to->x), FROM_FT_COORD(MINUS(to->y)));
+    return 0;
+}
+
+VPath VFont::outline(FT_UInt glyphIndex)
+{
+    auto search = mMap.find(glyphIndex);
```

Comment:
Font DB is thread safe but VFont is not .. multiple thread can call outline() function at the same time.